File: src/textbox.c
#include <SDL2/SDL_ttf.h>
#include <SDL2/SDL.h>

#include "util.h"
#include "border.h"
#include "colour.h"

#include "textbox.h"

float textbox_line_spacing = 1;
int textbox_vert_margin = 16;
int textbox_hori_margin = 16;

static SDL_Surface *textbox_draw_line(textbox *t, int lineno);
static SDL_Surface *textbox_draw_seg(
    char **text, col_rgb col, text_attrs attrs, TTF_Font *f
);
static void textbox_process_text_escapes(
    char **text, col_rgb *col, text_attrs *attrs
);
static SDL_Surface *textbox_join_surf_horiz(SDL_Surface *l, SDL_Surface *r);

int textbox_init(
    textbox *t, border *b, TTF_Font *font, SDL_Surface *surf, SDL_Rect *area
)
{
    t->font = font;
    t->surf = surf;
    t->bord = b;
    t->scroll = 0;

    textbox_resize(t, area);

    return 0;
}

void textbox_resize(textbox *t, SDL_Rect *newsize)
{
    float lnspacing;
    memcpy(&(t->area), newsize, sizeof(SDL_Rect));

    lnspacing = (float)TTF_FontHeight(t->font) * textbox_line_spacing;
    t->lnspacing = (int)round(lnspacing);

    if (t->attrs & no_margin)
        memcpy(&(t->textarea), &(t->area.x), sizeof(SDL_Rect));
    else
    {
        t->textarea.x = max(0, t->area.x + textbox_hori_margin);
        t->textarea.y = max(0, t->area.y + textbox_vert_margin);
        t->textarea.w = max(0, t->area.w - textbox_hori_margin * 2);
        t->textarea.h = max(0, t->area.h - textbox_vert_margin * 2);
    }

    t->vislines = max(1, t->textarea.h / t->lnspacing);
}

void textbox_draw(textbox *t)
{
    int linepos;

    if ( ! (t->attrs & no_border) )
        border_draw(t->bord, t->surf, &(t->area), t->attrs);

    for (linepos = 0; linepos < t->vislines; linepos++)
    {
        SDL_Rect linearea, dstrect;
        SDL_Surface *line;

        dstrect.x = t->textarea.x;
        dstrect.y = t->textarea.y + linepos * t->lnspacing;

        if (linepos + t->scroll >= t->numlines) break;

        line = textbox_draw_line(t, linepos + t->scroll);

        linearea.x = 0;
        linearea.y = 0;
        linearea.w = min(t->textarea.w, line->w);
        linearea.h = line->h;

        SDL_BlitSurface(line, &linearea, t->surf, &dstrect);

        SDL_FreeSurface(line);
    }
}

static void textbox_process_text_escapes(
    char **text, col_rgb *col, text_attrs *attrs
)
{
    if (*text == NULL || (*text)[0] == '\0')
    {
        *text = NULL;
        return;
    }

    if ((*text)[0] != '\\') return;

    switch ((*text)[1])
    {
    case 'c':
        if (!str_check_len(*text, 8))
            *text = NULL;
        else
            sscanf(*text, "\\c%2hhx%2hhx%2hhx", &(col->r), &(col->g), &(col->b));
        break;
    }
}

static SDL_Surface *textbox_draw_seg(
    char **text, col_rgb col, text_attrs attrs, TTF_Font *f
)
{
    SDL_Surface *rtn;
    char *nextseg;

    nextseg = strchr(*text, '\\');

    if (nextseg) *nextseg = '\0';
    rtn = TTF_RenderUTF8_Blended(f, *text, colour_to_sdl(col));
    if (nextseg) *nextseg = '\\';

    *text = nextseg;

    return rtn;
}


static SDL_Surface *textbox_join_surf_horiz(SDL_Surface *l, SDL_Surface *r)
{
    SDL_Surface *rtn;
    SDL_Rect     dstrect;

    rtn = surface_alpha(l->w + r->w, max(l->h, r->h));

    SDL_SetSurfaceBlendMode(l, SDL_BLENDMODE_NONE);
    SDL_SetSurfaceBlendMode(r, SDL_BLENDMODE_NONE);

    dstrect.y = 0;
    dstrect.x = 0;
    SDL_BlitSurface(l, NULL, rtn, &dstrect);
    dstrect.x = l->w;
    SDL_BlitSurface(r, NULL, rtn, &dstrect);

    SDL_FreeSurface(l);
    SDL_FreeSurface(r);

    return rtn;
}

static SDL_Surface *textbox_draw_line(textbox *t, int lineno)
{
    char *text;
    SDL_Surface *rtn, *seg;
    col_rgb col;

    rtn = NULL;

    col_set(col, 255, 255, 255);

    if (lineno >= t->numlines)
        return NULL;

    while (text)
    {
        textbox_process_text_escapes(&text, &col, NULL);
        if (!text) break;

        seg = textbox_draw_seg(&text, col, 0, t->font);
        if (!seg) break;

        if (rtn) rtn = textbox_join_surf_horiz(rtn, seg);
        else     rtn = seg;
    }

    return rtn;
}
/*
static SDL_Surface *textbox_draw_line(textbox *t, int lineno)
{
    SDL_Color col;
    char *text;
    char *dup, *seg, *cur;
    SDL_Surface *textseg, *rtn, *new;

    dup = strdup(t->text[lineno]);
    cur = dup;
    seg = strsep(&cur, "\\");
    rtn = NULL;

    col.r = 255;
    col.g = 255;
    col.b = 255;

    while (seg)
    {
        int width;
        textseg = TTF_RenderUTF8_Blended(t->font, seg, col);
        if (rtn)
        {
            SDL_Rect dstrect;
            new = SDL_CreateRGBSurface(
                0, textseg->w + rtn->w, textseg->h, 32, 0xff, 0xff00, 0xff0000, 0xff000000
            );
            SDL_FillRect(new, NULL, SDL_MapRGBA(new->format, 0, 0, 0, 0));

            dstrect.x = dstrect.y = 0;
            dstrect.h = textseg->h;

            dstrect.w = rtn->w;

            SDL_SetSurfaceBlendMode(rtn, SDL_BLENDMODE_NONE);
            SDL_BlitSurface(rtn, NULL, new, &dstrect);

            dstrect.w = textseg->w;
            dstrect.x = rtn->w;

            SDL_SetSurfaceBlendMode(textseg, SDL_BLENDMODE_NONE);
            SDL_BlitSurface(textseg, NULL, new, &dstrect);

            SDL_FreeSurface(rtn);
            SDL_FreeSurface(textseg);
            rtn = new;
        }
        else
        {
            rtn = textseg;
        }

        seg = strsep(&cur, "\\");
        
        if (seg)
            switch (*seg)
            {
            case 'c':
                printf("%s\n", seg);
                sscanf(seg, "c%2hhx%2hhx%2hhx", &(col.r), &(col.g), &(col.b));
                seg += 7;
                break;
            }
    }
    SDL_SetSurfaceBlendMode(rtn, SDL_BLENDMODE_BLEND);
    return rtn;
}

int main(void)
{
    border  b;
    textbox t;
    SDL_Window   *w;
    SDL_Renderer *r;
    SDL_Surface  *s;
    TTF_Font     *f;

    SDL_Init(SDL_INIT_VIDEO);
    TTF_Init();
    w = SDL_CreateWindow(
        "",
        SDL_WINDOWPOS_UNDEFINED,
        SDL_WINDOWPOS_UNDEFINED,
        513,
        513,
        0
    );

    if (w == NULL) {
        printf("Could not create window: %s\n", SDL_GetError());
        return 1;
    }

    r = SDL_CreateRenderer(w, -1, 0);

    if (r == NULL) {
        printf("Could not create renderer: %s\n", SDL_GetError());
        return 1;
    }

    s = SDL_GetWindowSurface(w);

    if (s == NULL) {
        printf("Could not create surface: %s\n", SDL_GetError());
        return 1;
    }

    f = TTF_OpenFont("/home/xeom/dev/rpg/asset/font/LiberationMono-Regular.ttf", 12);
    border_init(&b, "/home/xeom/dev/rpg/asset/art/borders/stone_brick.bmp");
    SDL_Rect area;

    area.x = 0;
    area.y = 0;
    area.w = 128;
    area.h = 128;
    textbox_init(&t, &b, f, s, &area);
    char *text[] = {"Hello,", "\\cff0000W\\cff7700o\\cffff00r\\c00ff00l\\c0000ffd"};
    t.numlines = 2;
    t.text = text;
    textbox_draw(&t);
    SDL_UpdateWindowSurface(w);
    SDL_Delay(2000);

    SDL_DestroyWindow(w);
    SDL_Quit();
    return 0;
}
*/

File: python/textbox.py
from so      import *
from sdl     import *
from border  import *
from display import *

class textbox_s(Structure):
    _fields_ = [
        ("bord",      border_p),
        ("font",      c_void_p),
        ("surf",      c_void_p),
        ("bordrrndr", c_void_p),
        ("area",      sdl_rect_s),
        ("textarea",  sdl_rect_s),
        ("vislines",  c_size_t),
        ("scroll",    c_size_t),
        ("attrs",     c_int),
        ("lnspacing", c_int),
        ("numlines",  c_size_t),
        ("text",      POINTER(c_char_p)),
    ]

textbox_p = POINTER(textbox_s)

so.textbox_init.argtypes = [textbox_p, border_p, c_void_p, c_void_p, c_int]
so.textbox_init.restype  = c_int

so.textbox_resize.argtypes = [c_void_p, c_void_p]
so.textbox_resize.restype  = None

so.textbox_draw.argtypes = [c_void_p]
so.textbox_draw.restype  = None

so.textbox_add_line.argtypes = [c_void_p, c_size_t]
so.textbox_add_line.restype  = None

so.textbox_set_line.argtypes = [c_void_p, c_size_t, c_char_p]
so.textbox_set_line.restype  = None

class Textbox:
    def __init__(self, ptr=None):
        if ptr == None:
            self.struct = textbox_s();
            self.ptr    = textbox_p(self.struct)
            self.ptr = cast(so.malloc(sizeof(textbox_s)), textbox_p)
            self.tofree = True

        else:
            self.tofree = False
            self.struct = ptr.contents
            self.ptr = ptr

    def __del__(self):
        pass
#        if self.tofree:
 #           so.free(self.ptr)

    def init(self, border, font, surf, attrs):
        so.textbox_init(self.ptr, border, font, surf, attrs)

    def resize(self, x, y, w, h):
        rect = sdl_rect_s(x, y, w, h)
        so.textbox_resize(self.ptr, byref(rect))

    def draw(self):
        so.textbox_draw(self.ptr)

    def add_line(self, i=None):
        so.textbox_add_line(self.ptr, i)

    def del_line(self, i):
        so.textbox_add_line(self.ptr, i)

    def set_line(self, i, text):
        print(i, text)
        so.textbox_set_line(self.ptr, i, text.encode("utf-8"))

class Menu:
    border_default = "wooden"
    font_default   = ("slkscr", 16)

    def __init__(self, surf, attrs, border=None, font=None):
        bordername = self.border_default if border == None else border
        fontname   = self.font_default   if font   == None else font

        self.border = Border.get(bordername)
        self.font   = Font.get(*fontname)

        self.textbox  = Textbox()

        self.textbox.init(self.border, self.font, surf, attrs)
        self.options  = []
        self.selected = None

        self.undrawn  = True

    def add_option(self, text):
        self.textbox.add_line(len(self.options))
        self.options.append(text)

        self.undrawn = True

        self.update_line(len(self.options) - 1)

        return len(self.options) - 1

    def update_line(self, i):
        if i == self.selected:
            self.textbox.set_line(i, "\\c4a3e0b \\cdf9a1b" + self.options[i])

        else:
            self.textbox.set_line(i, "\\c4a3e0b" + self.options[i])

        self.undrawn = True

    def select_option(self, i):
        prev = self.selected

        self.selected = i % len(self.options)

        if prev != None:
            self.update_line(prev)


        self.update_line(self.selected)

    def draw(self):
        self.textbox.draw()

    def on_option_choose(self, option):
        pass

    def on_option_select(self, option):
        pass

so.display_init()

m = Menu(cast(so.display_surface, POINTER(c_void_p)).contents, 15+128)
m.add_option("One")
m.add_option("Two")
m.add_option("Three")
m.add_option("Three")

m.add_option("Three")
m.add_option("Three")
m.add_option("Three")
m.add_option("Three")

def scroll(key):
    if key == "Up":
        if m.selected == None:
            m.select_option(1)
        m.select_option(m.selected - 1)

    if key == "Down":
        if m.selected == None:
            m.select_option(-1)
        m.select_option(m.selected + 1)

on_keydown.append(scroll)

import time

while cast(so.display_alive, POINTER(c_int)).contents.value:
    so.display_poll()
    m.draw()
    time.sleep(0.01)

#f = so.TTF_OpenFont(b"/home/xeom/dev/rpg/asset/font/LiberationMono-Bold.ttf", 12)
#b = border_s()
#b2 = border_s()
#
#so.border_init(byref(b), b"/home/xeom/dev/rpg/asset/art/borders/wooden.bmp")
#so.border_init(byref(b2), b"/home/xeom/dev/rpg/asset/art/borders/stone_brick.bmp")
#t = Textbox()
#t2 = Textbox()
#t2.init(b2, f, cast(so.display_surface, POINTER(c_void_p)).contents, 15)
#t.init(b, f, cast(so.display_surface, POINTER(c_void_p)).contents, 15)
#t.resize(16, 16, 128, 128)
#t.add_line(0)
#t.add_line(1)
#t.add_line(2)
#t.add_line(3)
#t.add_line(4)
#t.set_line(0, b"\c4a3e0bHello")
#t.set_line(1, b"\c4a3e0bHello")
#t.set_line(2, b"\cdf9a1b- World -")
#t.set_line(3, b"\c4a3e0bHello")
#t.set_line(4, b"\cd5c57bx Howdy x")
#t2.draw()
#t.draw()
#so.display_poll()
#import time
#time.sleep(10)
